pkg/hubble/exporter/api.go

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package exporter
+
+import (
+	"reflect"
+	"sort"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	"github.com/cilium/cilium/pkg/hubble/observer/observeroption"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// FlowLogExporter exports hubble events to configured target file.
+type FlowLogExporter interface {
+	observeroption.OnDecodedEvent
+	// Stop stops this exporter instance from further events processing.
+	Stop() error
+}
+
+// FlowFilters is a slice of filters to include or exclude flows.
+type FlowFilters []*flowpb.FlowFilter
+
+// FieldMask is a slice of fields that are included in output.
+type FieldMask []string
+
+// FlowLogConfig represents configuration of single dynamic exporter.
+type FlowLogConfig struct {
+	Name           string      `json:"name,omitempty" yaml:"name,omitempty"`
+	FilePath       string      `json:"filePath,omitempty" yaml:"filePath,omitempty"`
+	FieldMask      FieldMask   `json:"fieldMask,omitempty" yaml:"fieldMask,omitempty"`
+	IncludeFilters FlowFilters `json:"includeFilters,omitempty" yaml:"includeFilters,omitempty"`
+	ExcludeFilters FlowFilters `json:"excludeFilters,omitempty" yaml:"excludeFilters,omitempty"`
+	End            *time.Time  `json:"end,omitempty" yaml:"end,omitempty"`
+}
+
+func (f *FlowLogConfig) equals(other *FlowLogConfig) bool {
+	if f == nil && other == nil {
+		return true
+	}
+	if f == nil || other == nil {
+		return false
+	}
+
+	if f.FilePath != other.FilePath {
+		return false
+	}
+
+	if !f.FieldMask.equals(other.FieldMask) {
+		return false
+	}
+
+	if !f.IncludeFilters.equals(other.IncludeFilters) {
+		return false
+	}
+
+	if !f.ExcludeFilters.equals(other.ExcludeFilters) {
+		return false
+	}
+
+	if f.End == nil && other.End != nil ||
+		f.End != nil && other.End == nil ||
+		f.End != nil && other.End != nil && !f.End.Equal(*other.End) {
+		return false
+	}
+
+	return true
+}
+
+func (f FlowFilters) equals(other FlowFilters) bool {
+	aFiltersSet, bFiltersSet := make(map[string]bool), make(map[string]bool)
```

Comment:
Could you add a test case to verify if `protobuf.String()` will handle default values correctly? i.e. will the following flow protobufs will emit the same string:
```
flowpb.Flow {
  IP: nil,
}
```
vs.
```
flowpb.Flow {
  IP: &flowpb.IP{},
}
```

The perfect way will be comparing each other using `proto.Equal(aFilter, bFilter)`.

Reply:
The behavior is consistent with `proto.Equal` with one difference, that current solution ignores order in slices - which is desirable by this use case. I'm adding test case to cover this.